protocol/binary/stream_reader.go

```diff
@@ -0,0 +1,445 @@
+// Copyright (c) 2021 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package binary
+
+import (
+	"bytes"
+	"io"
+	"io/ioutil"
+	"math"
+
+	"go.uber.org/thriftrw/internal/iface"
+	"go.uber.org/thriftrw/protocol/stream"
+	"go.uber.org/thriftrw/wire"
+)
+
+// Requests for byte slices longer than this will use a dynamically resizing
+// buffer.
+const bytesAllocThreshold = 1048576 // 1 MB
+
+// For the reader, we keep track of the read offset manually everywhere so
+// that we can implement lazy collections without extra allocations
+
+// fixedWidth returns the encoded size of a value of the given type. If the
+// type's width depends on the value, -1 is returned.
+func fixedWidth(t wire.Type) int64 {
+	switch t {
+	case wire.TBool:
+		return 1
+	case wire.TI8:
+		return 1
+	case wire.TDouble:
+		return 8
+	case wire.TI16:
+		return 2
+	case wire.TI32:
+		return 4
+	case wire.TI64:
+		return 8
+	default:
+		return -1
+	}
+}
+
+// StreamReader provides an implementation of a "stream.Reader".
+type StreamReader struct {
+	iface.Private
+
+	reader io.Reader
+	buffer [8]byte
+}
+
+// NewStreamReader returns a new StreamReader.
+func NewStreamReader(r io.Reader) StreamReader {
+	return StreamReader{reader: r}
+}
+
+func (sr *StreamReader) read(bs []byte) (int, error) {
+	n, err := sr.reader.Read(bs)
+
+	// For consistency with the non-streaming Reader, return an
+	// io.ErrUnexpectedEOF if number of bytes read is smaller than what was
+	// expected, https://golang.org/pkg/io/#ReaderAt
+	if err == io.EOF || n < len(bs) {
+		// All EOFs are unexpected when streaming
+		err = io.ErrUnexpectedEOF
```

Comment:
Should we use https://golang.org/pkg/io/#ReadFull here or make our own version of it with a `for` loop?
Otherwise this won't handle non-EOF partial reads correctly.

Reply:
TIL about `ReadFull`, thanks - integrated that in.